compiler/rustc_lint/src/builtin.rs

```diff
@@ -3307,6 +3308,95 @@ impl EarlyLintPass for SpecialModuleName {
     }
 }
 
+declare_lint! {
+    /// The `ref_binop_on_copy_type` lint detects borrowed `Copy` types being passed to binary
+    /// operations that have unnecessary borrows.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// # #![allow(unused)]
+    /// pub fn slice_of_ints(input: &[(usize, usize, usize, usize)]) -> usize {
+    ///     input
+    ///         .iter()
+    ///         .filter(|(a, b, c, d)| a <= c && d <= b || c <= a && b <= d)
+    ///         .count()
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// When making comparisons (or other binary operations) between two reference values that can
+    /// be copied instead, the compiler will not always remove the references, making the execution
+    /// of the binary operation slower than it otherwise could be by making the machine code "chase
+    /// pointers" before actually finding the underlying value. If the `Copy` value is as small or
+    /// smaller than a 64 bit pointer, we suggest dereferencing the value so the compiler will have
+    /// a better chance of producing optimal instructions.
+    REF_BINOP_ON_COPY_TYPE,
+    Deny,
+    "detects binary operations on references to `Copy` types like `&42 < &50`",
+}
+
+declare_lint_pass!(RefBinopOnCopyType => [REF_BINOP_ON_COPY_TYPE]);
+
+impl<'tcx> LateLintPass<'tcx> for RefBinopOnCopyType {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'tcx>) {
+        let hir::ExprKind::Binary(op, left, right) = expr.kind else { return; };
+        let left_ty = cx.typeck_results().expr_ty_adjusted(left);
+        let right_ty = cx.typeck_results().expr_ty_adjusted(right);
+        if let Some(copy_trait) = cx.tcx.lang_items().copy_trait()
+            && let ty::Ref(_, left_ty, _) = left_ty.kind()
+            && let ty::Ref(_, left_ty, _) = left_ty.kind()
+            && let Some(_) = cx.tcx.non_blanket_impls_for_ty(copy_trait, *left_ty).next()
```

Comment:
Should we use `ty.is_copy_modulo_regions` perhaps?

Reply:
Yes